File: app/src/main/java/sgtmelon/scriptum/app/view/activity/IntroActivity.java
package sgtmelon.scriptum.app.view.activity;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.viewpager.widget.ViewPager;
import sgtmelon.scriptum.R;
import sgtmelon.scriptum.app.adapter.PagerAdapter;
import sgtmelon.scriptum.app.view.fragment.IntroFragment;
import sgtmelon.scriptum.office.annot.IntroAnn;
import sgtmelon.scriptum.office.st.PageSt;

public final class IntroActivity extends AppCompatActivity
        implements ViewPager.OnPageChangeListener {

    private static final String TAG = IntroActivity.class.getSimpleName();

    private final PagerAdapter pagerAdapter = new PagerAdapter(getSupportFragmentManager());;

    private View pageIndicator;
    private Button pageButtonEnd;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        Log.i(TAG, "onCreate");
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_intro);

        setupViewPager();
    }

    private void setupViewPager() {
        Log.i(TAG, "setupViewPager");

        ViewPager viewPager = findViewById(R.id.intro_pager);

        pageIndicator = findViewById(R.id.page_indicator);
        pageButtonEnd = findViewById(R.id.end_button);

        pageButtonEnd.setOnClickListener(view -> {
            Intent intent = new Intent(IntroActivity.this, MainActivity.class);
            startActivity(intent);
            finish();
        });

        pageButtonEnd.setAlpha(0);
        pageButtonEnd.setEnabled(false);

        IntroFragment introFragment;
        for (int i = 0; i < IntroAnn.count; i++) {
            introFragment = new IntroFragment();
            introFragment.setPageSt(new PageSt(i));
            pagerAdapter.addItem(introFragment);
        }

        viewPager.setAdapter(pagerAdapter);
        viewPager.setOffscreenPageLimit(pagerAdapter.getCount() - 1);
        viewPager.addOnPageChangeListener(this);
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        float targetAlpha = (float) Math.max(0.2, 1 - positionOffset);
        float targetScale = (float) Math.max(0.75, 1 - positionOffset);

        pagerAdapter.getItem(position).setChange(targetAlpha, targetScale);

        if (position != pagerAdapter.getCount() - 1) {
            targetAlpha = (float) Math.max(0.2, positionOffset);
            targetScale = (float) Math.max(0.75, positionOffset);

            pagerAdapter.getItem(position + 1).setChange(targetAlpha, targetScale);
        }

        if (position == pagerAdapter.getCount() - 1) {
            pageButtonEnd.setEnabled(true);
        } else pageButtonEnd.setEnabled(false);

        if (position == pagerAdapter.getCount() - 2) {
            pageIndicator.setTranslationY(positionOffset * pageButtonEnd.getHeight());
            pageIndicator.setAlpha(1 - positionOffset);
            pageButtonEnd.setAlpha(positionOffset);
        }
    }

    @Override
    public void onPageSelected(int position) {

    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }

}

File: app/src/main/java/sgtmelon/scriptum/app/view/activity/IntroActivity.kt
package sgtmelon.scriptum.app.view.activity

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import androidx.viewpager.widget.ViewPager
import sgtmelon.scriptum.R
import sgtmelon.scriptum.app.adapter.PagerAdapter
import sgtmelon.scriptum.app.view.fragment.IntroFragment
import sgtmelon.scriptum.office.data.IntroData
import sgtmelon.scriptum.office.state.PageState
import sgtmelon.scriptum.office.utils.AppUtils.beforeFinish

class IntroActivity : AppCompatActivity(), ViewPager.OnPageChangeListener {

    private val pagerAdapter = PagerAdapter(supportFragmentManager)

    private val pageIndicator: View by lazy {
        findViewById<View>(R.id.intro_page_indicator)
    }
    private val pageButtonEnd: Button by lazy {
        findViewById<Button>(R.id.intro_end_button)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_intro)

        setupViewPager()
    }

    private fun setupViewPager() {
        pageButtonEnd.setOnClickListener {
            beforeFinish {
                startActivity(Intent(this@IntroActivity, MainActivity::class.java))
            }
        }

        pageButtonEnd.alpha = 0f
        pageButtonEnd.isEnabled = false

        for (i in 0 until IntroData.count) {
            pagerAdapter.addItem(IntroFragment.getInstance(PageState(i)))
        }

        val viewPager = findViewById<ViewPager>(R.id.intro_pager)
        viewPager.adapter = pagerAdapter
        viewPager.offscreenPageLimit = pagerAdapter.count - 1
        viewPager.addOnPageChangeListener(this)
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        var targetAlpha = Math.max(0.2, (1 - positionOffset).toDouble()).toFloat()
        var targetScale = Math.max(0.75, (1 - positionOffset).toDouble()).toFloat()

        pagerAdapter.getItem(position).setChange(targetAlpha, targetScale)

        if (position != pagerAdapter.count - 1) {
            targetAlpha = Math.max(0.2, positionOffset.toDouble()).toFloat()
            targetScale = Math.max(0.75, positionOffset.toDouble()).toFloat()

            pagerAdapter.getItem(position + 1).setChange(targetAlpha, targetScale)
        }

        pageButtonEnd.isEnabled = position == pagerAdapter.count - 1

        if (position == pagerAdapter.count - 2) {
            pageIndicator.translationY = positionOffset * pageButtonEnd.height
            pageIndicator.alpha = 1 - positionOffset
            pageButtonEnd.alpha = positionOffset
        }
    }

    override fun onPageSelected(position: Int) {}

    override fun onPageScrollStateChanged(state: Int) {}

}